src/frontend/src/optimizer/plan_node/logical_join.rs

```diff
@@ -44,28 +44,44 @@ pub struct LogicalJoin {
     right: PlanRef,
     on: Condition,
     join_type: JoinType,
+    output_indices: Vec<usize>,
 }
 
 impl fmt::Display for LogicalJoin {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         write!(
             f,
-            "LogicalJoin {{ type: {:?}, on: {} }}",
-            &self.join_type, &self.on
+            "LogicalJoin {{ type: {:?}, on: {}, output_indices: {:?} }}",
+            &self.join_type,
+            &self.on,
+            &self.output_indices[..]
         )
     }
 }
 
 impl LogicalJoin {
     pub(crate) fn new(left: PlanRef, right: PlanRef, join_type: JoinType, on: Condition) -> Self {
+        let out_column_num =
+            Self::out_column_num(left.schema().len(), right.schema().len(), join_type);
+        Self::new_with_reordered_output(left, right, join_type, on, (0..out_column_num).collect())
+    }
+
+    pub(crate) fn new_with_reordered_output(
```

Comment:
`output_indices` can not only handle reorder but also column removals. Maybe `new_with_output_indices`?